libs/libvtrutil/src/vtr_thread_pool.h

```diff
@@ -0,0 +1,171 @@
+        size_t thread_idx = (next_thread++) % threads.size();
+        auto thread_data = threads[thread_idx].get();
+        size_t task_id = ++total_tasks_queued;
+
+        // VTR_LOG("Scheduling task %zu to thread %zu\n", task_id, thread_data->thread_id);
+
+        // Wrap the work with task completion tracking
+        auto task = [this, f = std::forward<F>(f), thread_id = thread_data->thread_id, task_id]() {
+            vtr::Timer task_timer;
+            // VTR_LOG("Thread %zu starting task %zu\n", thread_id, task_id);
+            
+            try {
+                f();
+                // VTR_LOG("Thread %zu completed task %zu successfully in %.3f seconds\n", 
+                //        thread_id, task_id, task_timer.elapsed_sec());
+            } catch (const std::exception& e) {
+                VTR_LOG_ERROR("Thread %zu failed task %zu with error: %s\n", 
+                             thread_id, task_id, e.what());
+                throw;
+            } catch (...) {
+                VTR_LOG_ERROR("Thread %zu failed task %zu with unknown error\n", 
+                             thread_id, task_id);
+                throw;
+            }
+
+            // Track task completion
+            size_t remaining = --active_tasks;
+            if (remaining == 0) {
+                completion_cv.notify_all();
+            }
+        };
+
+        // Queue the task
+        {
+            std::lock_guard<std::mutex> lock(thread_data->queue_mutex);
+            thread_data->task_queue.push(std::move(task));
+            // VTR_LOG("Task %zu queued to thread %zu (queue size: %zu)\n", 
+            //        task_id, thread_data->thread_id, thread_data->task_queue.size());
+        }
+        thread_data->cv.notify_one();
+    }
+
+    void wait_for_all() {
+        std::unique_lock<std::mutex> lock(completion_mutex);
+        completion_cv.wait(lock, [this]() { return active_tasks == 0; });
+    }
+
+    ~thread_pool() {
+        // VTR_LOG("Shutting down thread pool after %.3f seconds, processed %zu total tasks\n", 
+        //        pool_timer.elapsed_sec(), total_tasks_queued.load());
+        
+        // Signal all threads to stop
+        for (auto& thread_data : threads) {
+            {
+                std::lock_guard<std::mutex> lock(thread_data->queue_mutex);
+                thread_data->stop = true;
+                // VTR_LOG("Signaling thread %zu to stop (remaining tasks: %zu)\n", 
+                //        thread_data->thread_id, thread_data->task_queue.size());
+            }
+            thread_data->cv.notify_one();
+        }
+
+        // Join all threads
+        for (auto& thread_data : threads) {
+            if (thread_data->thread.joinable()) {
+                thread_data->thread.join();
+                // VTR_LOG("Thread %zu joined after completing %zu tasks\n", 
+                //        thread_data->thread_id, thread_data->tasks_completed);
+            }
+        }
+    }
+};
+
+}  // namespace vtr
```

Comment:
Also you tell me you have some dead variables / experimental code to delete here. If you no longer need it, it would be good to delete before the PR merge.